libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,598 @@
+        }
+    }
+
+    // If we haven't heard from a vehicle then assume it is no threat
+    if (sample_age > MAX_SAMPLE_AGE_MS) {
+        sample.threat_level = MAV_COLLISION_THREAT_LEVEL_NONE;
+    }
+
+    // could optimise this to not calculate a lot of this if threat
+    // level is none - but only *once the GCS has been informed*!
+    sample.closest_approach_xy = closest;
+    float current_distance = get_distance(my_loc, sample_loc);
+    sample.distance_to_closest_approach = current_distance - closest;
+    Vector2f net_velocity_ne = Vector2f(my_vel[0] - sample_vel[0], my_vel[1] - sample_vel[1]);
+    sample.time_to_closest_approach = 0.0f;
+    if (!is_zero(sample.distance_to_closest_approach) &&
+        ! is_zero(net_velocity_ne.length())) {
+        sample.time_to_closest_approach = sample.distance_to_closest_approach / net_velocity_ne.length();
+    }
+}
+
+void AP_Avoidance::handle_threat_gcs_notify(AP_Avoidance::Sample *threat)
+{
+    if (threat == nullptr) {
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
+    if (threat->threat_level == MAV_COLLISION_THREAT_LEVEL_NONE) {
+        // only send cleared messages for a few seconds:
+        if (_gcs_cleared_messages_first_sent == 0) {
+            _gcs_cleared_messages_first_sent = now;
+        }
+        if (now - _gcs_cleared_messages_first_sent > _gcs_cleared_messages_duration * 1000) {
+            return;
+        }
+    } else {
+        _gcs_cleared_messages_first_sent = 0;
+    }
+    if (now - threat->last_gcs_report_time > _gcs_notify_interval * 1000) {
+        GCS_MAVLINK::send_collision_all(*threat, mav_avoidance_action());
+        threat->last_gcs_report_time = now;
+    }
+
+}
+
+void AP_Avoidance::check_for_threats()
+{
+    Location my_loc;
+    if (!_ahrs.get_position(my_loc)) {
+        // if we don't know our own location we can't determine any threat level
+        return;
+    }
+
+    Vector3f my_vel;
+    if (!_ahrs.get_velocity_NED(my_vel)) {
+        // assuming our own velocity to be zero here may cause us to
+        // fly into something.  Better not to attempt to avoid in this
+        // case.
+        return;
+    }
+
+    // we always check all samples to see if they are threats since it
+    // is most likely our own position and/or velocity have changed
+    // determine the current most-serious-threat
+    _current_most_serious_threat = -1;
+    for (uint8_t i=0; i<_sample_count; i++) {
+
+        AP_Avoidance::Sample &sample = _samples[i];
+        const uint32_t sample_age = AP_HAL::millis() - sample.timestamp_ms;
+        debug("i=%d src_id=%d timestamp=%u age=%d\n", i, sample.src_id, sample.timestamp_ms, sample_age);
+
+        update_threat_level(my_loc, my_vel, sample);
+        debug("   threat-level=%d", sample.threat_level);
+
+        // ignore any really old data:
+        if (sample_age > MAX_SAMPLE_AGE_MS) {
+            continue;
+        }
+
+        bool sample_is_more_serious_threat = false;
+        do {
```

Comment:
Beats the mess you'd have with a big-ass "if" statement IMO :-)

Moved it into a function.